script/add_gpl_license.py

```diff
@@ -0,0 +1,184 @@
+# contributors
+#
+# https://github.com/vegastrike/Vega-Strike-Engine-Source
+#
+# This file is part of Vega Strike.
+#
+# Vega Strike is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Vega Strike is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Vega Strike.  If not, see <https://www.gnu.org/licenses/>.
+
+"""This script takes a file path (or paths) as an argument, and rewrites the file
+with the Vega Strike GPL 3 copyright notice prepended to it.
+
+If no arguments are passed, the script will read from stdin.
+
+The Copyright text is copied from the LICENSE file, which is assumed to be at
+the root of the current git repository. Hence this script needs to be run from
+within a git repository containing a LICENSE file at the root. Lines
+containing the text 'Copyright (C)' at the start of the LICENSE file will be
+copied as the Copyright text for the GPL notice. e.g. If the LICENSE file
+contains the following lines, they will be copied in at the start of the GPL
+notice:
+
+    Copyright (c) 2001-2002 Daniel Horn
+    Copyright (c) 2002-2019 pyramid3d and other Vega Strike Contributors
+    Copyright (c) 2019-2021 Stephen G. Tuggy, and other Vega Strike Contributors
+
+Usage:
+    python3 add_gpl_license.py [-h|--help]
+    python3 add_gpl_license.py PATH [PATH ...]
+"""
+
+import sys
+from pathlib import Path
+from shutil import copystat, move
+from tempfile import NamedTemporaryFile
+
+LICENSE_TEXT = """{filename}
+
+{copyright_notice}
+
+https://github.com/vegastrike/Vega-Strike-Engine-Source
+
+This file is part of Vega Strike.
+
+Vega Strike is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Vega Strike is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Vega Strike. If not, see <https://www.gnu.org/licenses/>."""
+
+# Characters to use for the start, middle and end of a comment block
+C_LIKE_COMMENT = ('/**', '*', '*/')
+SCRIPT_LIKE_COMMENT = ('##', '#', '#')
+
+COMMENTS_BY_FILE_SUFFIX = {
+    '.c': C_LIKE_COMMENT,
+    '.cpp': C_LIKE_COMMENT,
+    '.h': C_LIKE_COMMENT,
+    '.hpp': C_LIKE_COMMENT,
+    '.h.in': C_LIKE_COMMENT,
```

Comment:
I don't think `'.h.in'` should be included here. Other than that, things are looking pretty good so far.

Reply:
Oh, I think I remember. This file is input for CMake, which uses it to generate config.h. So this file does belong in the repository, and does need a copyleft notice. So never mind.

Reply:
I was using ripgrep's filetype filters to get the list of files to update. `.h.in` and `.txt` are both there for CMake files.